File: nodejs/lib/type-format.js
/** Format Cloud Spanner google.spanner.v1.Type values. */

import {
  TYPE_ANNOTATION_NAMES,
  TypeAnnotationCode,
  TypeCode,
  typeCodeName,
} from './codes.js';
import { UnknownTypeError } from './errors.js';
import {
  arrayElementType,
  fieldName,
  fieldType,
  protoTypeFqn,
  structFields,
  typeAnnotation,
  typeCode,
} from './proto.js';

export const StructMode = Object.freeze({
  BASE: 0,
  RECURSIVE: 1,
  RECURSIVE_WITH_NAME: 2,
});

export const ProtoEnumMode = Object.freeze({
  BASE: 0,
  LEAF: 1,
  FULL: 2,
  LEAF_WITH_KIND: 3,
  FULL_WITH_KIND: 4,
});

export const ArrayMode = Object.freeze({
  BASE: 0,
  RECURSIVE: 1,
});

export const UnknownMode = Object.freeze({
  UNKNOWN: 0,
  TYPE_CODE: 1,
  VERBOSE: 2,
  PANIC: 3,
});

export const TypeAnnotationMode = Object.freeze({
  SUFFIX: 0,
  OMIT: 1,
  PRIMARY: 2,
});

/** @typedef {Object} FormatOption
 * @property {number} [struct]
 * @property {number} [proto]
 * @property {number} [enum]
 * @property {number} [array]
 * @property {number} [unknown]
 * @property {number} [typeAnnotation]
 */

export const FORMAT_OPTION_SIMPLEST = Object.freeze({
  struct: StructMode.BASE,
  proto: ProtoEnumMode.BASE,
  enum: ProtoEnumMode.BASE,
  array: ArrayMode.BASE,
  unknown: UnknownMode.TYPE_CODE,
  typeAnnotation: TypeAnnotationMode.SUFFIX,
});

export const FORMAT_OPTION_SIMPLE = Object.freeze({
  struct: StructMode.BASE,
  proto: ProtoEnumMode.LEAF,
  enum: ProtoEnumMode.LEAF,
  array: ArrayMode.RECURSIVE,
  unknown: UnknownMode.UNKNOWN,
  typeAnnotation: TypeAnnotationMode.SUFFIX,
});

export const FORMAT_OPTION_NORMAL = Object.freeze({
  struct: StructMode.RECURSIVE,
  proto: ProtoEnumMode.LEAF,
  enum: ProtoEnumMode.LEAF,
  array: ArrayMode.RECURSIVE,
  unknown: UnknownMode.VERBOSE,
  typeAnnotation: TypeAnnotationMode.SUFFIX,
});

export const FORMAT_OPTION_VERBOSE = Object.freeze({
  struct: StructMode.RECURSIVE_WITH_NAME,
  proto: ProtoEnumMode.FULL,
  enum: ProtoEnumMode.FULL,
  array: ArrayMode.RECURSIVE,
  unknown: UnknownMode.VERBOSE,
  typeAnnotation: TypeAnnotationMode.SUFFIX,
});

export const FORMAT_OPTION_MORE_VERBOSE = Object.freeze({
  struct: StructMode.RECURSIVE_WITH_NAME,
  proto: ProtoEnumMode.FULL_WITH_KIND,
  enum: ProtoEnumMode.FULL_WITH_KIND,
  array: ArrayMode.RECURSIVE,
  unknown: UnknownMode.VERBOSE,
  typeAnnotation: TypeAnnotationMode.SUFFIX,
});

/**
 * @param {string} s
 * @param {string} sep
 */
function lastCut(s, sep) {
  const idx = s.lastIndexOf(sep);
  return idx >= 0 ? s.slice(idx + sep.length) : s;
}

/**
 * @param {number} ann
 */
function annotationSuffix(ann) {
  if (ann === TypeAnnotationCode.TYPE_ANNOTATION_CODE_UNSPECIFIED) {
    return '';
  }
  const name = TYPE_ANNOTATION_NAMES[ann];
  if (name == null) {
    return `(${ann})`;
  }
  return `(${name})`;
}

/**
 * @param {number} ann
 */
function annotationName(ann) {
  const name = TYPE_ANNOTATION_NAMES[ann];
  if (name == null) {
    return String(ann);
  }
  return name;
}

/**
 * @param {number} code
 * @param {number} [mode]
 */
export function formatTypeCode(code, mode = UnknownMode.VERBOSE) {
  const name = typeCodeName(code);
  if (name != null) {
    return name;
  }
  if (mode === UnknownMode.TYPE_CODE) {
    return String(code);
  }
  if (mode === UnknownMode.VERBOSE) {
    return `UNKNOWN(${code})`;
  }
  if (mode === UnknownMode.PANIC) {
    throw new UnknownTypeError(`unknown TypeCode(${code})`);
  }
  return 'UNKNOWN';
}

/**
 * @param {unknown} typ
 * @param {number} mode
 */
export function formatProtoEnum(typ, mode) {
  const code = typeCode(typ);
  const fqn = protoTypeFqn(typ);
  const codeName = formatTypeCode(code);
  if (mode === ProtoEnumMode.LEAF) {
    return lastCut(fqn, '.');
  }
  if (mode === ProtoEnumMode.FULL) {
    return fqn;
  }
  if (mode === ProtoEnumMode.LEAF_WITH_KIND) {
    return `${codeName}<${lastCut(fqn, '.')}>`;
  }
  if (mode === ProtoEnumMode.FULL_WITH_KIND) {
    return `${codeName}<${fqn}>`;
  }
  return codeName;
}

/**
 * @param {unknown[]} fields
 * @param {FormatOption} option
 */
export function formatStructFields(fields, option) {
  const parts = [];
  for (const field of fields) {
    const typeStr = formatType(fieldType(field), option);
    if (option.struct === StructMode.RECURSIVE_WITH_NAME && fieldName(field)) {
      parts.push(`${fieldName(field)} ${typeStr}`);
    } else {
      parts.push(typeStr);
    }
  }
  return parts.join(', ');
}

/**
 * @param {unknown} typ
 * @param {FormatOption} option
 */
function formatTypeImpl(typ, option) {
  const code = typeCode(typ);
  if (code === TypeCode.ARRAY && option.array !== ArrayMode.BASE) {
    const elem = arrayElementType(typ);
    return `ARRAY<${formatType(elem, option)}>`;
  }
  if (code === TypeCode.PROTO) {
    return formatProtoEnum(typ, option.proto ?? ProtoEnumMode.BASE);
  }
  if (code === TypeCode.ENUM) {
    return formatProtoEnum(typ, option.enum ?? ProtoEnumMode.BASE);
  }
  if (code === TypeCode.STRUCT && option.struct !== StructMode.BASE) {
    return `STRUCT<${formatStructFields(structFields(typ), option)}>`;
  }
  return formatTypeCode(code, option.unknown ?? UnknownMode.VERBOSE);
}

/**
 * Format a Cloud Spanner Type as a string.
 * @param {unknown} typ
 * @param {FormatOption | null | undefined} [option]
 */
export function formatType(typ, option = FORMAT_OPTION_SIMPLE) {
  const opt = option ?? FORMAT_OPTION_SIMPLE;
  const ann = typeAnnotation(typ);
  const annMode = opt.typeAnnotation ?? TypeAnnotationMode.SUFFIX;
  if (annMode === TypeAnnotationMode.OMIT) {
    return formatTypeImpl(typ, opt);
  }
  if (annMode === TypeAnnotationMode.PRIMARY) {
    if (ann !== TypeAnnotationCode.TYPE_ANNOTATION_CODE_UNSPECIFIED) {
      return annotationName(ann);
    }
    return formatTypeImpl(typ, opt);
  }
  return formatTypeImpl(typ, opt) + annotationSuffix(ann);
}

/** @param {unknown} typ */
export function formatTypeSimplest(typ) {
  return formatType(typ, FORMAT_OPTION_SIMPLEST);
}

/** @param {unknown} typ */
export function formatTypeSimple(typ) {
  return formatType(typ, FORMAT_OPTION_SIMPLE);
}

/** @param {unknown} typ */
export function formatTypeNormal(typ) {
  return formatType(typ, FORMAT_OPTION_NORMAL);
}

/** @param {unknown} typ */
export function formatTypeVerbose(typ) {
  return formatType(typ, FORMAT_OPTION_VERBOSE);
}

/** @param {unknown} typ */
export function formatTypeMoreVerbose(typ) {
  return formatType(typ, FORMAT_OPTION_MORE_VERBOSE);
}

/** @param {unknown} typ */
export function formatTypeVerboseAnnotationOmit(typ) {
  return formatType(typ, {
    ...FORMAT_OPTION_VERBOSE,
    typeAnnotation: TypeAnnotationMode.OMIT,
  });
}

/** @param {unknown} typ */
export function formatTypeVerboseAnnotationPrimary(typ) {
  return formatType(typ, {
    ...FORMAT_OPTION_VERBOSE,
    typeAnnotation: TypeAnnotationMode.PRIMARY,
  });
}

File: rust/src/type_format.rs
//! Format Cloud Spanner `google.spanner.v1.Type` values.

use crate::codes::{
    type_annotation_name, type_code_name, TypeAnnotationCode, TypeCode,
};
use crate::errors::{FormatError, Result, UnknownTypeError};
use crate::types::{array_element_type, field_name, proto_type_fqn, struct_fields, Type};

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum StructMode {
    Base = 0,
    Recursive = 1,
    RecursiveWithName = 2,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ProtoEnumMode {
    Base = 0,
    Leaf = 1,
    Full = 2,
    LeafWithKind = 3,
    FullWithKind = 4,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum ArrayMode {
    Base = 0,
    Recursive = 1,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum UnknownMode {
    Unknown = 0,
    TypeCode = 1,
    Verbose = 2,
    Panic = 3,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub enum TypeAnnotationMode {
    Suffix = 0,
    Omit = 1,
    Primary = 2,
}

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
pub struct FormatOption {
    pub struct_mode: StructMode,
    pub proto: ProtoEnumMode,
    pub enum_mode: ProtoEnumMode,
    pub array: ArrayMode,
    pub unknown: UnknownMode,
    pub type_annotation: TypeAnnotationMode,
}

pub const FORMAT_OPTION_SIMPLEST: FormatOption = FormatOption {
    struct_mode: StructMode::Base,
    proto: ProtoEnumMode::Base,
    enum_mode: ProtoEnumMode::Base,
    array: ArrayMode::Base,
    unknown: UnknownMode::TypeCode,
    type_annotation: TypeAnnotationMode::Suffix,
};

pub const FORMAT_OPTION_SIMPLE: FormatOption = FormatOption {
    struct_mode: StructMode::Base,
    proto: ProtoEnumMode::Leaf,
    enum_mode: ProtoEnumMode::Leaf,
    array: ArrayMode::Recursive,
    unknown: UnknownMode::Unknown,
    type_annotation: TypeAnnotationMode::Suffix,
};

pub const FORMAT_OPTION_NORMAL: FormatOption = FormatOption {
    struct_mode: StructMode::Recursive,
    proto: ProtoEnumMode::Leaf,
    enum_mode: ProtoEnumMode::Leaf,
    array: ArrayMode::Recursive,
    unknown: UnknownMode::Verbose,
    type_annotation: TypeAnnotationMode::Suffix,
};

pub const FORMAT_OPTION_VERBOSE: FormatOption = FormatOption {
    struct_mode: StructMode::RecursiveWithName,
    proto: ProtoEnumMode::Full,
    enum_mode: ProtoEnumMode::Full,
    array: ArrayMode::Recursive,
    unknown: UnknownMode::Verbose,
    type_annotation: TypeAnnotationMode::Suffix,
};

pub const FORMAT_OPTION_MORE_VERBOSE: FormatOption = FormatOption {
    struct_mode: StructMode::RecursiveWithName,
    proto: ProtoEnumMode::FullWithKind,
    enum_mode: ProtoEnumMode::FullWithKind,
    array: ArrayMode::Recursive,
    unknown: UnknownMode::Verbose,
    type_annotation: TypeAnnotationMode::Suffix,
};

fn last_cut(s: &str, sep: char) -> &str {
    s.rsplit_once(sep).map(|(_, tail)| tail).unwrap_or(s)
}

fn annotation_suffix(ann: i32) -> String {
    if ann == TypeAnnotationCode::TypeAnnotationCodeUnspecified as i32 {
        return String::new();
    }
    match type_annotation_name(ann) {
        Some(name) => format!("({name})"),
        None => format!("({ann})"),
    }
}

fn annotation_name(ann: i32) -> String {
    type_annotation_name(ann)
        .map(str::to_string)
        .unwrap_or_else(|| ann.to_string())
}

pub fn format_type_code(code: i32, mode: UnknownMode) -> Result<String> {
    if let Some(name) = type_code_name(code) {
        return Ok(name.to_string());
    }
    match mode {
        UnknownMode::TypeCode => Ok(code.to_string()),
        UnknownMode::Verbose => Ok(format!("UNKNOWN({code})")),
        UnknownMode::Panic => Err(FormatError::UnknownType(UnknownTypeError::new(format!(
            "unknown TypeCode({code})"
        )))),
        UnknownMode::Unknown => Ok("UNKNOWN".to_string()),
    }
}

pub fn format_proto_enum(typ: &Type, mode: ProtoEnumMode) -> String {
    let code = typ.code;
    let fqn = proto_type_fqn(typ);
    let code_name = type_code_name(code).unwrap_or("UNKNOWN");
    match mode {
        ProtoEnumMode::Leaf => last_cut(fqn, '.').to_string(),
        ProtoEnumMode::Full => fqn.to_string(),
        ProtoEnumMode::LeafWithKind => format!("{code_name}<{}>", last_cut(fqn, '.')),
        ProtoEnumMode::FullWithKind => format!("{code_name}<{fqn}>"),
        ProtoEnumMode::Base => code_name.to_string(),
    }
}

pub fn format_struct_fields(fields: &[crate::types::Field], option: FormatOption) -> Result<String> {
    let mut parts = Vec::with_capacity(fields.len());
    for field in fields {
        let type_str = format_type(&field.field_type, Some(option))?;
        if option.struct_mode == StructMode::RecursiveWithName && !field_name(field).is_empty() {
            parts.push(format!("{} {type_str}", field_name(field)));
        } else {
            parts.push(type_str);
        }
    }
    Ok(parts.join(", "))
}

fn format_type_impl(typ: &Type, option: FormatOption) -> Result<String> {
    let code = typ.code;
    if code == TypeCode::Array as i32 && option.array != ArrayMode::Base {
        let elem = array_element_type(typ)
            .ok_or_else(|| FormatError::UnknownType(UnknownTypeError::new("ARRAY missing element type")))?;
        return Ok(format!("ARRAY<{}>", format_type(elem, Some(option))?));
    }
    if code == TypeCode::Proto as i32 {
        return Ok(format_proto_enum(typ, option.proto));
    }
    if code == TypeCode::Enum as i32 {
        return Ok(format_proto_enum(typ, option.enum_mode));
    }
    if code == TypeCode::Struct as i32 && option.struct_mode != StructMode::Base {
        return Ok(format!(
            "STRUCT<{}>",
            format_struct_fields(struct_fields(typ), option)?
        ));
    }
    format_type_code(code, option.unknown)
}

pub fn format_type(typ: &Type, option: Option<FormatOption>) -> Result<String> {
    let option = option.unwrap_or(FORMAT_OPTION_SIMPLE);
    let ann = typ.type_annotation;
    match option.type_annotation {
        TypeAnnotationMode::Omit => format_type_impl(typ, option),
        TypeAnnotationMode::Primary => {
            if ann != TypeAnnotationCode::TypeAnnotationCodeUnspecified as i32 {
                Ok(annotation_name(ann))
            } else {
                format_type_impl(typ, option)
            }
        }
        TypeAnnotationMode::Suffix => {
            let mut out = format_type_impl(typ, option)?;
            out.push_str(&annotation_suffix(ann));
            Ok(out)
        }
    }
}

pub fn format_type_simplest(typ: &Type) -> Result<String> {
    format_type(typ, Some(FORMAT_OPTION_SIMPLEST))
}

pub fn format_type_simple(typ: &Type) -> Result<String> {
    format_type(typ, Some(FORMAT_OPTION_SIMPLE))
}

pub fn format_type_normal(typ: &Type) -> Result<String> {
    format_type(typ, Some(FORMAT_OPTION_NORMAL))
}

pub fn format_type_verbose(typ: &Type) -> Result<String> {
    format_type(typ, Some(FORMAT_OPTION_VERBOSE))
}

pub fn format_type_more_verbose(typ: &Type) -> Result<String> {
    format_type(typ, Some(FORMAT_OPTION_MORE_VERBOSE))
}

pub fn format_type_verbose_annotation_omit(typ: &Type) -> Result<String> {
    format_type(
        typ,
        Some(FormatOption {
            type_annotation: TypeAnnotationMode::Omit,
            ..FORMAT_OPTION_VERBOSE
        }),
    )
}

pub fn format_type_verbose_annotation_primary(typ: &Type) -> Result<String> {
    format_type(
        typ,
        Some(FormatOption {
            type_annotation: TypeAnnotationMode::Primary,
            ..FORMAT_OPTION_VERBOSE
        }),
    )
}
